### ISSUE TITLE
Bug: Request?: bind all class methods by default.

### ISSUE THREAD

**Issue by NullVoxPopuli** [statelyai/xstate#3024]
### Description

It's a long standing problem in JavaScript that classes' methods are unbound by default. It's kinda bonkers, but it's where we're at.

The react folks, back when they were using classes, made all methods arrow functions, which are _properties who's value happens to be an arrow function_, but they inherit the class instance's `this`, which is the desired behavior.

```js
class Foo {
  myMethod = () => { /* `this` is correct here */ }
}
```

The ember folks made a decorator which binds the this, so that the semantics of real methods could be used.
```js
class Foo {
  @action myMethod() {
    /* `tihs` is correct here */
  }
}
```

This only comes up when functions are passed as referenced to other places to be called "by something".

For example, on the `State` class, there is this `can` method:
https://github.com/statelyai/xstate/blob/main/packages/core/src/State.ts#L359

If you pass that method elsewhere to invoke later, the `this` is lost.

A reduced example: https://jsbin.com/tojigos/edit?html,output

```js
class State {
  foo = 'my state';
      
  can() {
    return this.foo;
  }
}
    
let state = new State();
    
console.log(state.can());
    
function evaluateFn(fn) {
  // error: this is undefined
  return fn();
}
    
let result = evaluateFn(state.can);

console.log({ result }); // never reached
```

### Expected result

that I can pass function references around and have their context maintained -- this requires the `this` to be explicitly bound.

### Actual result

Error: `this` is undefined.

### Reproduction

https://jsbin.com/tojigos/edit?html,output

### Additional context

This was initially reported here: https://github.com/NullVoxPopuli/ember-functions-as-helper-polyfill/issues/103

**Comment by Andarist** [comment]
I'm not super keen on making such a change for all methods etc. This has some memory implications (each instance gets its own copies of those "methods") and at a framework/library level this can be a concern. There are obvious workarounds for this - either binding or creating a wrapper function that captures the `state`. 

I see how this is a PITA for methods that are often used in render functions though so I think that we could introduce such a change for `state.can`. Note that `.bind`ing can be better for this purpose than using the class fields syntax - at least it was better (JS engine-wise) a couple of years back, I didn't recheck this with the new versions of V8 and other engines.

**Comment by NullVoxPopuli** [comment]
I put up a PR: https://github.com/statelyai/xstate/pull/3061

**Comment by Andarist** [comment]
We decided against binding those methods. This is somewhat unfortunate because destructuring those methods won't work. However, we gain simplicity, better performance, and decreased memory footprint thanks to that. Those are treated as methods by us and this is how `this` works in JavaScript.

**Comment by NullVoxPopuli** [comment]
How much of an impact is this to memory?
Isn't this type of object only created a handful of times per page? I'd expect impact to be unnoticeable

**Comment by Andarist** [comment]
On a page? Yeah, you might not have a lot of those objects. In a game? That might already become a bigger issue.

But it's not only about the performance. It's also about how we can leverage this to integrate with other frameworks. Notice how I was able to just remove this custom logic that was effectively trying to "fight" the binding etc: https://github.com/statelyai/xstate/pull/4447/files/7c1497318799c6abe638e3dfa4639f86b880c18f#diff-ae937e88f3bc909d18c32bfa82f6ee16160971751a06671cf0e094264609c79d

I understand why binding might be desirable. It's just that it also has its tradeoffs and we decided to choose the tradeoffs of not binding those methods. Any layer that decides to integrate XState with X can always just decide to bind those and that will be a fine choice. 

We currently can't actually do this but I'm actively thinking about ways to narrow down the context type based on `.matches` call etc. This is challenging to pull off in TS and will definitely require typegen if we ever pull it off. But if we do... it will only work with `state.matches`. In that potential future, destructuring `matches` out of `state` and using it standalone will lose all of its narrowing capabilities in TS (it's a hard-ish limitation on the TS side of things). So I just think that it makes sense to keep passing around `state` as a whole object.
